ublox_derive/src/output.rs

```diff
@@ -0,0 +1,562 @@
+                pub fn #get_name(&self) -> #ty {
+                    #(#get_value_lines)*
+                }
+            });
+            off += size_bytes;
+        } else {
+            assert_eq!(field_index, pack_descr.fields.len() - 1);
+            let mut get_value_lines = vec![quote! { &self.0[#off..] }];
+
+            if let Some(ref out_ty) = f.map.map_type {
+                let get_raw = &get_value_lines[0];
+                let new_line = quote! { let val = #get_raw ;  };
+                get_value_lines[0] = new_line;
+                let from_fn = &out_ty.from_fn;
+                get_value_lines.push(quote! {
+                    #from_fn(val)
+                });
+
+                if f.map.convert_may_fail {
+                    let is_valid_fn = &out_ty.is_valid_fn;
+                    field_validators.push(quote! {
+                        let val = &payload[#off..];
+                        if !#is_valid_fn(val) {
+                            return Err(ParserError::InvalidField{
+                                packet: #pack_name,
+                                field: stringify!(#get_name)
+                            });
+                        }
+                    });
+                }
+            }
+            let out_ty = if f.has_intermidiate_type() {
+                ty.clone()
+            } else {
+                parse_quote! { &[u8] }
+            };
+            getters.push(quote! {
+                #[doc = #field_comment]
+                #[inline]
+                pub fn #get_name(&self) -> #out_ty {
+                    #(#get_value_lines)*
+                }
+            });
+        }
+    }
+    let struct_comment = &pack_descr.comment;
+    let validator = if let Some(payload_len) = pack_descr.packet_payload_size() {
+        quote! {
+            fn validate(payload: &[u8]) -> Result<(), ParserError> {
+                let expect = #payload_len;
+                let got = payload.len();
+                if got ==  expect {
+                    #(#field_validators)*
+                    Ok(())
+                } else {
+                    Err(ParserError::InvalidPacketLen{ packet: #pack_name, expect, got })
+                }
+            }
+        }
+    } else {
+        let min_size = pack_descr
+            .packet_payload_size_except_last_field()
+            .expect("except last all fields should have fixed size");
+
+        quote! {
+            fn validate(payload: &[u8]) -> Result<(), ParserError> {
+                let min = #min_size;
+                let got = payload.len();
+                if got >= min {
+                    #(#field_validators)*
+                    Ok(())
+                } else {
+                    Err(ParserError::InvalidPacketLen{ packet: #pack_name, expect: min, got })
+                }
+            }
+        }
+    };
+
+    quote! {
+        #[doc = #struct_comment]
+        #[doc = "It is just reference to internal parser's buffer"]
```

Comment:
> Conceivably, an end user could use one of these to reference their own internal parser buffer, right?

No, definition looks like this `struct TestRef<'a>(&'a [u8])` (from ublox_derive/src/tests.rs), as you can see content of `TestRef` is private, so user can not construct it by (him|her)self  in his/her  "internal parser".

Because of `PacketRef` is designed in invariant that it created against "valid buffer",
and our parser can guarantee this, while about other "internal parser" it is absolutely unknown.